python/lsst/daf/butler/core/dimensions/dataId.py

```diff
@@ -0,0 +1,351 @@
+    hold additional metadata, regions, and definitions for those `Dimension`\s.
+    They are also hashable, and hence can be used as keys in dictionaries.
+
+    The `DataId` class represents a complete ID that has either been obtained
+    from or validated with the set of known `Dimension`\s.  Regular `dict`\s
+    are typically used to represent conceptual data IDs that have not been
+    validated.
+    The `DataId` constructor serves as a sort of standardization routine; most
+    APIs that conceptually operate on `DataId`\s should accept either true
+    `DataId`\s or regular dicts via a single ``dataId`` argument, and pass
+    this through the `DataId` construction (usually with additional keyword
+    arguments forwarded) to guarantee a true `DataId`.  When convenient, that
+    `DataId` should also be returned.
+
+    The set of dimensions a `DataId` identifies can be provided to the
+    constructor four ways:
+
+    - Multiple dimensions may be passed via the ``dimensions`` argument.
+    - A single dimension may be passed via the ``dimension`` argument.
+    - If a true `DataId` is passed, its dimensions will be used if they are
+      not overridden by one of the above.
+    - If none of the above is provided, the dimensions are inferred from the
+      set of keys provided in ``dataId``, ``extra``, and ``kwds``; any
+      dimensions in ``universe`` whose links are a subset of those keys is
+      included.
+
+    Parameters
+    ----------
+    dataId : `dict` or `DataId`
+        A `dict`-like object containing `Dimension` links.  If this is a true
+        `DataId` and the set of dimensions identified does not change, this
+        object will be updated in-place and returned instead of a new instance.
+    dimensions : iterable of `Dimension` or `str`, optional
+        The set of dimensions the `DataId` will identify, either as `Dimension`
+        instances or string names thereof.
+    dimension : `Dimension` or `str`, optional
+        The single dimension this `DataId` will identify (along with all of
+        its required dependencies).
+    universe : `DimensionGraph`, optional
+        A graph containing all known dimensions and joins.  Must be provided
+        if names are passed instead of `Dimension` instances in ``dimensions``
+        or ``dimension``, or when dimensions are inferred from the provided
+        link keys.
+    region : `lsst.sphgeom.ConvexPolygon`, optional
+        Spatial region on the sky associated with this combination of
+        dimension entries.
+    entries : `dict`, optional
+        A nested dictionary of additional metadata column values associated
+        with these dimensions, with `DimensionElement` instances or `str`
+        names as the outer keys, `str` column names as inner keys, and
+        column values as innter dictionary values.
+    extra : `dict`, optional
+        Additional key-value pairs to update ``dataId`` with.
+    kwds : `dict`, optional
```

Comment:
```rst
kwds
    Additional keyword arguments passed to `TODO`.
```

Where `TODO` is the API object, or objects, that receive the keyword arguments.

Note that there's no type information on the `kwds` parameter either. To do the user, kwds isn't a `dict`.